File: src/main/java/com/comma/newsbot/service/NewsService.java
package com.comma.newsbot.service;

import com.comma.newsbot.domain.News;
import com.comma.newsbot.fetcher.RssFetcher;
import com.comma.newsbot.repository.AvailableFeedsRepository;
import com.comma.newsbot.repository.NewsRepository;
import com.comma.newsbot.service.channel.Discord;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Random;

@Service
@Slf4j
public class NewsService {

    private final AvailableFeedsRepository feedsRepository;
    private final Discord discord;
    private final NewsRepository newsRepository;
    private final Random random = new Random();
    private final RssFetcher rssFetcher;

    public NewsService(AvailableFeedsRepository feedsRepository, Discord discord,
                       NewsRepository newsRepository, RssFetcher rssFetcher){
        this.feedsRepository = feedsRepository;
        this.discord = discord;
        this.newsRepository = newsRepository;
        this.rssFetcher = rssFetcher;
    }

    public News chooseNews(){
        try {
            String url = pickRandomUrl();
            List<News> listNews = rssFetcher.fetchNews(url);

            for(News news : listNews){
                if(!newsRepository.existsByUrl(news.getUrl())){
                    discord.sendMessage(news);
                    newsRepository.save(news);
                    return news;
                }
            }
            log.info("There isn't a new news available in this source.");
        } catch(Exception e){
            log.error("Failed to choose news ", e);
        }
        return null;
    }

    private String pickRandomUrl(){
        List<String> activeUrls = feedsRepository.findAllActiveUrls();
        if(activeUrls.isEmpty()){
            return null;
        }

        int idx = random.nextInt(activeUrls.size());

        return activeUrls.get(idx);
    }

}

File: src/service/news_service.py
import logging
import random
from sqlalchemy.orm import Session
from src.fetcher.rss_fetcher import fetch_news
from src.service.llm.groq_llm import GroqLLM
import src.repository.news_repository as news_repo
import src.repository.available_feeds_repository as feeds_repo

logger = logging.getLogger(__name__)

_llm = GroqLLM()


def choose_news(db: Session, chat_id: str = None):
    try:
        active_urls = feeds_repo.find_all_active_urls(db)
        if not active_urls:
            logger.warning("No active feed URLs found.")
            return None

        url = random.choice(active_urls)
        news_list = fetch_news(url)

        for news in news_list:
            if not news_repo.exists_by_url(db, news.url):
                news.description = _llm.summarize_news(news.description)
                news_repo.save(db, news)

                if chat_id:
                    from src.service.channel.telegram import Telegram

                    Telegram().send_message(news, chat_id)

                return news

        logger.info("No new articles found in this source.")
    except Exception as e:
        logger.error(f"Failed to choose news: {e}")
    return None
